### PR TITLE
[SPARK-20935][STREAMING] Always close WriteAheadLog and make it idempotent

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to stop `ReceiverTracker` to close `WriteAheadLog` whenever it is and make `WriteAheadLog` and its implementations idempotent.

## How was this patch tested?

Added a test in `WriteAheadLogSuite`. Note that  the added test looks passing even if it closes twice (namely even without the changes in `FileBasedWriteAheadLog` and `BatchedWriteAheadLog`. It looks both are already idempotent but this is a rather sanity check.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @srowen and @zsxwing. Could you take a look and see if it makes sense?

File: streaming/src/main/java/org/apache/spark/streaming/util/WriteAheadLog.java [hunk 1]
```diff
@@ -56,7 +56,7 @@
   public abstract void clean(long threshTime, boolean waitForCompletion);
 
   /**
-   * Close this log and release any resources.
+   * Close this log and release any resources. It requires to be idempotent.
```
Comment (srowen) [inline streaming/src/main/java/org/apache/spark/streaming/util/WriteAheadLog.java]: "It must be idempotent" or "It is required to be idempotent"

File: streaming/src/main/scala/org/apache/spark/streaming/util/BatchedWriteAheadLog.scala [hunk 2]
```diff
@@ -122,6 +122,7 @@ private[util] class BatchedWriteAheadLog(val wrappedLog: WriteAheadLog, conf: Sp
   override def close(): Unit = {
     logInfo(s"BatchedWriteAheadLog shutting down at time: ${System.currentTimeMillis()}.")
     synchronized {
+      if (!active) return
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/util/BatchedWriteAheadLog.scala]: `active` might be simpler as an `AtomicBoolean`. Then I think you can just write `if (!active.getAndSet(false)) return` as the very first line.

File: streaming/src/main/scala/org/apache/spark/streaming/util/FileBasedWriteAheadLog.scala [hunk 3]
```diff
@@ -205,10 +205,12 @@ private[streaming] class FileBasedWriteAheadLog(
 
   /** Stop the manager, close any open log writer */
   def close(): Unit = synchronized {
-    if (currentLogWriter != null) {
-      currentLogWriter.close()
+    if (!executionContext.isShutdown) {
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/util/FileBasedWriteAheadLog.scala]: Does this need to be thread-safe? it seems like the impl just above tries to be. I don't know if it's needed.

Comment (HyukjinKwon) [inline streaming/src/main/scala/org/apache/spark/streaming/util/FileBasedWriteAheadLog.scala]: @srowen, maybe I missed your point. However, wouldn't there be a change to try `shutdown` multiple times if `isShutdown` is not synchronised together? Probably, `shutdown` is idempotent itself but just wonder if it is okay to keep this as a safeguard.

Comment (HyukjinKwon) [inline streaming/src/main/scala/org/apache/spark/streaming/util/FileBasedWriteAheadLog.scala]: Ah! do you actually mean checking if `shutdown` is idempotent. Let me check out.

Comment (HyukjinKwon) [inline streaming/src/main/scala/org/apache/spark/streaming/util/FileBasedWriteAheadLog.scala]: It looks okay to call multiple times.

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html#shutdown--

> Invocation has no additional effect if already shut down.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**